### DISCUSSION TITLE
Confusion on usage of programs from the example repo

### DISCUSSION THREAD

**Discussion by zamorae** [eModbus/eModbus#99] [Q&A]
**Hi eModbus team!** 
Thanks for your efforts in developing a Modbus solution for ESP32. I am new to this repository, but I have been looking for a reliable ESP32 modbus solution, and this seems to be what I need. I have setup a couple of 2 ESP32 devkitC devices interfaced to MAX3081 processors for RS-485 operation using Ethernet cable (CAT6e). 
In order to carry out initial tests, I want to configure one of the ESP32 as Master (Client) and the other as Slave (Server), and simply send a repetitive code from the Master to be seen on the slave, and a reply from the slave to the Master. From there I will proceed to increase the slave counts and increase the distance to verify the performances. 
My question at this point is: **which of the examples can be used for the Client and which one for the Server?** 
I see a list of 12 examples but I'm not sure how to implement for each of my test devices based off such list. From what I understand:

- All examples beginning with TCP are for WiFi, so for now I will focus on the wire based examples.
- WiFi-RTU bridge is a pass-through between wireless and non-wireless layers, will leave it aside for the moment.

That leaves "RTU06example" and "RTU16example" as Clients. **Can you please explain further on the difference between these?**
The "RTUserver" example seems to be the one to use for any server type of device, therefore one of the ESP32s will run either RTU06 or RTU16 and the other one used as slave must run the "RTUserver" example, **is this correct?**

**Can you explain in more detail what is the function of example "LoopbackClientServer"?** I see that the structure calls for WIFI Server and Client, does it mean the code must be adapted to Client or Server and load into **2 independent processors to test?** **can you provide more information on the function?**

Also, is the "Generate Message" example intended to be used as stand-alone in a single ESP32 to **simulate reception of messages and corresponding handling?**
 Basically any of the 2 "RTUXX" examples combined with the "RTUserver" example loaded in the different processors should be enough to generate messages/responses (chit chat) between devices, **correct?**

I have seen the documentation but it is still not very clear to me how should I load the programs in the independent processors to be able to see their corresponding outputs on the Serial Terminals. I know most people uses a temperature or humidity sensor that has its own code and interfaces with the programs, but my objective at this point is get the data flowing between 2 ESP32s, later I will move on to sensors and other RTU based devices.

Sorry if I'm being a bit obtuse on this topic, but I want to comprehend in detail the reasoning behind each of your valuable examples.

Thanks in advance!
-EZ

**Comment by Miq1** [comment]
A very first guess if a certain example is meant to work as a server or a client can be made by looking at the eModbus header file that is used within. The header files are named with "Client" or "Server" in their names.

That aside, there is no pair of examples intended to be used for both a client and a server. The closest one is the loopback example, that uses both server and client on a single ESP32, tied together by the use of the WiFi loopback function - hence the name. 

The difference between the RTU06 and RTU16 examples is the Modbus function code handled - it is 06 (write register) vs. 16 (write multiple registers). 

I would suggest setting up one of your ESP32 as a server serving just the 03 (read register) function code along the respective example, the other as a client requesting registers according a client example doing that and extending it once you got this working. 

eModbus is a library, not a set of ready-made sketches, because it never will know how a certain application is supposed to work. So I am afraid you will have to learn to use it and code your very task yourself...

**Reply by zamorae** [reply]
Thanks for the reply. I understand the fact of it being a library.... I just wondered if there was a direct relation Client-Server coded into the examples for non TCP operations to be used via two separate units, as it would seem normal (at least to me) to use at least one client / one server in the real world and was thinking this was the case with the examples. That is fine; I will check out the examples and modify them to suit the task and proceed from there.

Thanks again for the clarification points.
-EZ
